source/asynch_socket.c

```diff
@@ -533,12 +533,29 @@ static socket_error_t lwipv4_socket_bind(struct socket *sock, const struct socke
 }
 
 static void rx_core(struct socket * s, struct pbuf *p) {
-
+    /** NOTE: the receive callback for both UDP and TCP are responsible for deallocating pbufs, so there is no call to
+     *  pbuf_ref() required here.
+     */
     __disable_irq();
     if (s->rxBufChain == NULL) {
         s->rxBufChain = p;
     } else {
-        pbuf_cat((struct pbuf *) s->rxBufChain, p);
+        struct pbuf * q = (struct pbuf *)s->rxBufChain;
+        switch(s->family) {
+            case SOCKET_DGRAM:
+                // find the last element of the buffer chain.
+                while (q->next) {q = q->next;}
+                /**
+                 * Attach p to it wihout changing the tot_len of the buffer chain
+                 * NOTE: This is not how pbufs are intended to work, but it is necessary to deal with a) fragmentation
+                 * and b) packet queueing
+                 */
+                q->next = p;
```

Comment:
Are you sure this is enough though? lwIP has `pbuf_chain` and `pbuf_cat` and they seem to do a bit more than just setting `next`